packages/dds/tree/src/core/rebase/rebaser.ts

```diff
@@ -3,223 +3,139 @@
+					change,
+					parent: effectiveBase,
+				};
+				targetRebasePath.push({ ...c, change });
+			}
+			inverses.unshift(tagInverse(this.changeRebaser.invert(c), c.revision));
+		}
 
-	/**
-	 * Informs the Rebaser that `revision` will not be used again,
-	 * and internal resources related to it may be freed.
-	 */
-	public discardRevision(revision: RevisionTag): void {
-		throw Error("Not implemented"); // TODO
+		// Compose all changes together to get a single change that represents the entire rebase operation
+		return [effectiveBase, this.changeRebaser.compose([...inverses, ...targetRebasePath])];
 	}
-}
-
-// TODO: managing the types with this is not working well (inferring any for methods in Rebaser). Do something else.
-export type ChangesetFromChangeRebaser<TChangeRebaser extends ChangeRebaser<any>> =
-	TChangeRebaser extends ChangeRebaser<infer TChangeset> ? TChangeset : never;
-
-/**
- * Rebasing logic for a particular kind of change.
- *
- * This interface is used to provide rebase policy to `Rebaser`.
- *
- * The implementation must ensure TChangeset forms a [group](https://en.wikipedia.org/wiki/Group_(mathematics)) where:
- * - `compose([])` is the identity element.
- * - associativity is defined as `compose([...a, ...b])` is equal to
- * `compose([compose(a), compose(b)])` for all `a` and `b`.
- * - `inverse(a)` gives the inverse element of `a`.
- *
- * In these requirements the definition of equality is up to the implementer,
- * but it is required that any two changes which are considered equal:
- * - have the same impact when applied to any tree.
- * - can be substituted for each-other in all methods on this
- * interface and produce equal (by this same definition) results.
- *
- * For the sake of testability, implementations will likely want to have a concrete equality implementation.
- *
- * This API uses `compose` on arrays instead of an explicit identity element and associative binary operator
- * to allow the implementation more room for optimization,
- * but should otherwise be equivalent to the identity element and binary operator group approach.
- *
- * TODO:
- * Be more specific about the above requirements.
- * For example, would something that is close to forming a group but has precision issues
- * (ex: the floating point numbers and addition) be ok?
- * Would this cause decoherence (and thus be absolutely not ok),
- * or just minor semantic precision issues, which could be tolerated.
- * For now assume that such issues are not ok.
- *
- * @alpha
- */
-export interface ChangeRebaser<TChangeset> {
-	_typeCheck?: Invariant<TChangeset>;
 
 	/**
-	 * Compose a collection of changesets into a single one.
-	 * See {@link ChangeRebaser} for requirements.
+	 * Rebase a change over the given source and target branches
+	 * @param change - the change to rebase
+	 * @param source - the source branch to rebase backwards over
+	 * @param target - the target branch to rebase forwards over
+	 * @returns the rebased change
```

Comment:
I think it might be clearer to say that `source` is the branch/revision `change` is based on, and `target` is the branch/revision we want the change to be rebased onto.